capsules/src/seven_segment.rs

```diff
@@ -0,0 +1,397 @@
+//!   - `data1`: The position of the digit. Starts at 1.
+//!   - `data2`: The custom pattern to be represented.
+//!   - Return: `Ok(())` if the index was valid, `INVAL` otherwise.
+
+use core::cell::Cell;
+
+use kernel::hil::gpio::{ActivationMode, Pin};
+use kernel::hil::time::{Alarm, AlarmClient, ConvertTicks};
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::TakeCell;
+use kernel::ErrorCode;
+use kernel::ProcessId;
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::SevenSegment as usize;
+
+/// Digit patterns
+//
+//      A
+//      _
+//   F |_| B       center = G
+//   E |_| C . Dp
+//      D
+//
+const DIGITS: [u8; 10] = [
+    // pattern: 0bDpGFEDCBA
+    0b00111111, // 0
+    0b00000110, // 1
+    0b01011011, // 2
+    0b01001111, // 3
+    0b01100110, // 4
+    0b01101101, // 5
+    0b01111101, // 6
+    0b00100111, // 7
+    0b01111111, // 8
+    0b01101111, // 9
+];
+
+/// Holds an array of digits and an array of segments for each digit.
+
+pub struct SevenSegmentDriver<'a, P: Pin, A: Alarm<'a>, const NUM_DIGITS: usize> {
+    /// An array of 8 segment pins (7 for digit segments and one dot segment)
+    segments: &'a [&'a P; 8],
+    /// An array of `NUM_DIGITS` digit pins, each one corresponding to one digit on the display
+    /// For each digit selected, a pattern of lit and unlit segments will be represented
+    digits: &'a [&'a P; NUM_DIGITS],
+    /// A buffer which contains the patterns displayed for each digit
+    /// Each element of the buffer array represents the pattern for one digit, and
+    /// is a sequence of bits that have the value 1 for a lit segment and the value 0 for
+    /// an unlit segment.
+    buffer: TakeCell<'a, [u8; NUM_DIGITS]>,
+    alarm: &'a A,
+    current_digit: Cell<usize>,
+    /// How fast the driver should switch between digits (ms)
+    timing: u8,
+    segment_activation: ActivationMode,
+    digit_activation: ActivationMode,
+}
+
+impl<'a, P: Pin, A: Alarm<'a>, const NUM_DIGITS: usize> SevenSegmentDriver<'a, P, A, NUM_DIGITS> {
+    pub fn new(
+        segments: &'a [&'a P; 8],
+        digits: &'a [&'a P; NUM_DIGITS],
+        buffer: &'a mut [u8; NUM_DIGITS],
+        alarm: &'a A,
+        segment_activation: ActivationMode,
+        digit_activation: ActivationMode,
+        refresh_rate: usize,
+    ) -> Self {
+        // Check if the buffer has enough space to hold patterns for all digits
+        if (buffer.len() * 8) < segments.len() * digits.len() {
+            panic!("Digits Driver: provided buffer is too small");
+        }
+
+        Self {
+            segments,
+            digits,
+            buffer: TakeCell::new(buffer),
+            alarm,
+            segment_activation: segment_activation,
+            digit_activation: digit_activation,
+            current_digit: Cell::new(0),
+            timing: (1000 / (refresh_rate * digits.len())) as u8,
```

Comment:
Why not make `timing` a `u32` by default? The refresh rate is _usually_ high enough that `timing` should be less than 255 ms, but there's no guarantee of that in the equation here.

Reply:
You could go even further and just store `interval` instead of `timing`, since there's no option to change refresh rate anyways.